Add cache examples, update to latest actions

karlhorky opened the pull request:
Closes #2

karlhorky committed:
Add cache examples, update to latest actions

Closes #2

karlhorky commented on .github/workflows/nvm.yml:
```diff
@@ -5,9 +5,47 @@ jobs:
       - run: |
           nvm install --latest-npm --no-progress
-          echo "::set-env name=PATH::${NVM_BIN}:${PATH}"
+          echo "$(dirname $(which node))" >> $GITHUB_PATH
```
if this should stay as adding `NVM_BIN` to the `$PATH`, happy to also make that change - just wasn't sure whether the `node` binary was also available at `NVM_BIN`

ybiquitous replied:
Great catch! `NVM_BIN` is no longer necessary. 👍🏼

ybiquitous commented on .github/workflows/nvm.yml:
[suggest] Can you add `pull_request` to verify this PR?

https://github.com/ybiquitous/using-nvm-on-github-action/blob/4da2e4d46a2307e4d7c3083fbdc7142b6000b93e/.github/workflows/nvm.yml#L2

E.g.

```diff
-on: [push]
+on:
+  push: {branches: ["master"]}
+  pull_request:
```

karlhorky replied:
done in 1706c0d386af2a3d167981a3418c80cf2c7235b0

ybiquitous commented on .github/workflows/nvm.yml:
```diff
@@ -5,9 +5,47 @@ jobs:
       - run: |
           nvm install --latest-npm --no-progress
-          echo "::set-env name=PATH::${NVM_BIN}:${PATH}"
+          echo "$(dirname $(which node))" >> $GITHUB_PATH
```
[suggest] Let's use quotes for environment variables.

```suggestion
          echo "$(dirname $(which node))" >> "$GITHUB_PATH"
```

karlhorky replied:
Yeah normally I would do this but in the docs for GitHub's "magic" environment variables, they don't include quotes

So I would lean slightly towards not quoting them, to keep it close to the docs

ybiquitous replied:
> they don't include quotes

I didn't notice that, for example: https://docs.github.com/en/actions/sharing-automations/creating-actions/creating-a-composite-action#creating-an-action-metadata-file

Okay, let's keep them as-is. 👍🏼

ybiquitous commented on .github/workflows/nvm.yml:
```diff
@@ -5,9 +5,47 @@ jobs:
       - run: |
           nvm install --latest-npm --no-progress
-          echo "::set-env name=PATH::${NVM_BIN}:${PATH}"
+          echo "$(dirname $(which node))" >> $GITHUB_PATH
```
[suggest] I'm unsure but is `nvm which` safer than just `which`?

```suggestion
          echo "$(dirname $(nvm which node))" >> $GITHUB_PATH
```

karlhorky replied:
Oh interesting, I didn't know `nvm which` was a command - happy to apply if this works

karlhorky replied:
so according to [the docs](https://github.com/nvm-sh/nvm#:~:text=You%20can%20also%20get%20the%20path%20to%20the%20executable%20to%20where%20it%20was%20installed%3A) it appears that we will need a Node.js version specifier to get a path:

> You can also get the path to the executable to where it was installed:
>
> ```
> nvm which 12.22
> ```

if we want to go this direction, we could try using `--latest-npm` or `*` or something...?

...or maybe `nvm which` (without any arguments) also works because of the `.nvmrc` file?

karlhorky replied:
actually, looking closer at [the version aliases in the `nvm` docs](https://github.com/nvm-sh/nvm#:~:text=In%20place%20of,version%20of%20node), there is also a `node` version alias, weird!

> In place of a version pointer like "14.7" or "16.3" or "12.22.1", you can use the following special default aliases with `nvm install`, `nvm use`, `nvm run`, `nvm exec`, `nvm which`, etc:
>
>  - `node`: this installs the latest version of [`node`](https://nodejs.org/en/)

so I switched to `nvm which node` in 3f4af212a6b31e2107db034df429371cb82365f2

...although I do wonder whether this will conflict with the `14.4.0` version in [`.nvmrc`](https://github.com/ybiquitous/using-nvm-on-github-action/blob/master/.nvmrc) 🤔

ybiquitous replied:
`nvm which` is proper because `.nvmrc` version is used expectedly:

https://github.com/ybiquitous/using-nvm-on-github-action/pull/3/checks#step:6:7

<img width="211" alt="image" src="https://github.com/user-attachments/assets/58d1ba1e-c996-4a2a-9e33-928757809536">

https://github.com/ybiquitous/using-nvm-on-github-action/blob/8548472d5e23e281c073ed854bae95a3e5c67e87/.nvmrc#L1

karlhorky replied:
Great, so we'll leave `nvm which node` as it currently is? (since it's working)

Or switch to `nvm which`?

ybiquitous replied:
Let's leave `nvm which node` as-is 🙏🏼

ybiquitous commented on .github/workflows/nvm.yml:
```diff
@@ -5,9 +5,47 @@ jobs:
+      - name: Get npm cache directory
+        id: npm-cache-dir
+        shell: bash
+        run: echo "dir=$(npm config get cache)" >> $GITHUB_OUTPUT
```
```suggestion
        run: echo "dir=$(npm config get cache)" >> "$GITHUB_OUTPUT"
```

ybiquitous commented on .github/workflows/nvm.yml:
```diff
@@ -5,9 +5,47 @@ jobs:
+        id: npm-cache
+        with:
+          path: ${{ steps.npm-cache-dir.outputs.dir }}
+          key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.os }}-node-
```
[suggest] Is `npm-` more consistent than `node-`?

```suggestion
          key: ${{ runner.os }}-npm-${{ hashFiles('**/package-lock.json') }}
          restore-keys: |
            ${{ runner.os }}-npm-
```

ybiquitous commented on .github/workflows/nvm.yml:
```diff
@@ -5,9 +5,47 @@ jobs:
+      # Optional: Cache Yarn v1 dependencies
+      - name: Get yarn cache directory path
+        id: yarn-cache-dir-path
+        run: echo "dir=$(yarn cache dir)" >> $GITHUB_OUTPUT
```
```suggestion
        run: echo "dir=$(yarn cache dir)" >> "$GITHUB_OUTPUT"
```

ybiquitous commented on .github/workflows/nvm.yml:
```diff
@@ -5,9 +5,47 @@ jobs:
+            ${{ runner.os }}-node-
+
+      # Optional: Cache Yarn v1 dependencies
+      - name: Get yarn cache directory path
+        id: yarn-cache-dir-path
```
[suggest] Let's keep consistency with `Get npm cache directory` and `npm-cache-dir`.

```suggestion
      - name: Get yarn cache directory
        id: yarn-cache-dir
```

ybiquitous commented on .github/workflows/nvm.yml:
```diff
@@ -5,9 +5,47 @@ jobs:
+      - name: Get pnpm store directory
+        shell: bash
+        run: |
+          echo "STORE_PATH=$(pnpm store path --silent)" >> $GITHUB_ENV
```
[suggest] Can you use `$GITHUB_OUTPUT` also here for consistency?
I'm also worried that `STORE_PATH` may be conflicted with other environment variables.

karlhorky replied:
done in c3ba283cd0517ab5fd94fbfa7df95075b612baab

ybiquitous commented on .github/workflows/nvm.yml:
```diff
@@ -5,9 +5,47 @@ jobs:
+      - name: Get pnpm store directory
+        shell: bash
+        run: |
+          echo "STORE_PATH=$(pnpm store path --silent)" >> $GITHUB_ENV
```
[ask] Do we need to install `pnpm`? It doesn't seem to be included in the pre-installed software list:
https://github.com/actions/runner-images/blob/ubuntu22/20240804.1/images/ubuntu/Ubuntu2204-Readme.md

karlhorky replied:
Yes, if we want this action to be runnable out of the box

We can use pnpm/action-setup https://github.com/pnpm/action-setup

ybiquitous replied:
I didn't know that action. Let's include it!

karlhorky replied:
done in 219c6ee92f90a8618ecec7270167b7b2dd0d7da1

ybiquitous reviewed (changes requested):
@karlhorky Thanks for the pull request!

Overall LGTM, but I've left some reviews. Can you take a look?

karlhorky committed:
Fix cache key name

Co-authored-by: Masafumi Koba <473530+ybiquitous@users.noreply.github.com>

karlhorky committed:
Change names for consistency

Co-authored-by: Masafumi Koba <473530+ybiquitous@users.noreply.github.com>

karlhorky committed:
Enable running on PRs

karlhorky committed:
Add pnpm/action-setup

karlhorky committed:
Switch pnpm cache steps to use $GITHUB_OUTPUT

karlhorky committed:
Switch `which node` to `nvm which node`

From the docs: https://github.com/nvm-sh/nvm#:~:text=In%20place%20of,version%20of%20node

> In place of a version pointer like "14.7" or "16.3" or "12.22.1", you can use the following special default aliases with `nvm install`, `nvm use`, `nvm run`, `nvm exec`, `nvm which`, etc:
> 
>  - `node`: this installs the latest version of [`node`](https://nodejs.org/en/)

ybiquitous commented:
We have to update `.nvmrc` to resolve the pnpm error, right? 

<img width="528" alt="image" src="https://github.com/user-attachments/assets/3c89a752-e49e-4181-84c4-a6508ae34494">

https://github.com/ybiquitous/using-nvm-on-github-action/pull/3/checks#step:11:8

karlhorky committed:
Upgrade to Node.js 20.16.0

karlhorky committed:
Upgrade to pnpm v9

karlhorky commented:
> We have to update `.nvmrc` to resolve the pnpm error, right?

Maybe, pnpm v8 dropped support for Node.js v14:

- https://github.com/pnpm/pnpm/issues/6297#issuecomment-1486551881

I upgraded to v20.16.0 (latest LTS) in this commit:

- 05a3f7965e06a91aad68dea5674bc3e2a6556941

Since new versions of `pnpm@latest` will cause this to probably break in future again, I also pinned the pnpm version to v9:

- 7723a26cb09dbd880ee96c1cf24319b2930fe7ce

ybiquitous reviewed (approved):
Looks great! Thank you 😄 👍🏼

karlhorky commented:
Glad to help :) Thanks for the guidance, review and merge!
